lib/Canonicalize/MakeModelGEPPass.cpp

```diff
@@ -1908,10 +1834,10 @@ getAccessedTypeOnIR(FunctionMetadataCache &Cache,
   } break;
 
   default:
-    return std::nullopt;
+    return nullptr;
```

Comment:
I'm not a fan of this usage of `nullptr`. Can we return `model::UpcastableType()` or maybe `model::UpcastableType::empty()`?

Reply:
I feel like `nullptr` is a lot more expressive. But I guess I'd be fine with something like
```
class UpcastablePointer {
  constexpr std::nullptr_t nullopt = nullptr;
  // ...
};
```

Reply:
I'd go with `UpcastablePointer()`, or maybe `UpcastablePointer::empty()` but that likely already taken by the method *checking* if it's empty.

Reply:
To be honest, I really prefer things like this to be constexpr static objects instead of members.

But also, let's finish the discussion I started on mattermost first. Depending on how we decide to proceed with nullability, this problem might just go away 🤞.

Reply:
I'm fine with `constexpr` static objects. My main point is: users should never have to type `nullptr`. The fact that we use a pointer is an implementation detail.

Reply:
I agree that it's an implementation detail. But also it's a really important one. Because of how it's structured, they are forced to check it for _nullptr_ all over the place anyway (even if through the `empty()` helper I snatched from `TupleTreeReference`).

At one point, I seriously considered to do something like
```diff
-using UpcastableType = UpcastablePointer<Type>
+struct UpcastableType : UpcastablePointer<Type>, CommonTypeMethods<UpcastablePointer<Type>> {}; 
```

to alleviate that somewhat (since in practice, the `getX`/`toX`/`isX` things are used way more commonly than other stuff), but that brings more headaches than it solves. And, maybe, one day, we can outlaw `nullptr`-set `UpcastableType`s to cut the problem away in its root.  One can only dream.

Reply:
What's the final decision on this matter?

Reply:
```cpp
  constexpr static auto empty() noexcept { return nullptr; }
  bool isEmpty() const noexcept { return Pointer == empty(); }
```

Reply:
Done.